Add retryWhen()

blesh opened the pull request:
This addresses #482

A few things to note:

1. I *really* wanted to replicate prior art here, especially for the unit tests, however I found the unit tests for `retry` to be a bit confusing, and upon closer examination I realized they weren't being run at all. I'm not sure why that would be, so I've also emitted unit tests for `retryWhen`, at least for now.

2. I've added a `Enumerable.prototype.catchErrorWhen` in lieu of updating `Enumerable.prototype.catchError`. This is because it seems `catchError` serves a very specific, synchronous function, and altering it to use Observables broke a number of other tests.

In light of #1 ... I'll try to get at least some basic unit test coverage into this PR as soon as I can. I'm looking for feedback. I could easily add the tests I've kept locally, but I didn't want to submit something that was too radically different from prior art.

blesh commented:
Okay, I've added at least one test and it's now in rx, rx.all, and rx.lite.

mattpodwysocki commented:
@blesh there are a number of issues before I can accept this PR.
1. You must have deterministic tests using the `TestScheduler` which should be easy to do with `scheduler.createHotObservable` and `scheduler.createColdObservable`.
2. This must have associated documentation and be linked to [`Observable`](https://github.com/Reactive-Extensions/RxJS/blob/master/doc/api/core/observable.md)
3. This must also be associated with the appropriate files such as [`rx.lite.js`](https://github.com/Reactive-Extensions/RxJS/blob/master/doc/libraries/rx.lite.md) and [`rx.js`](https://github.com/Reactive-Extensions/RxJS/blob/master/doc/libraries/rx.md)
4.  You must remove any files from this commit in the dist folder as it makes it impossible to merge properly.

blesh commented:
:+1: I'll dig into that as soon as I can.

blesh commented:
Okay, after talking with @stealthcode about the RxJava implementation, I realized I had a few things wrong. So I've updated the RxJS implementation to reflect what is present in RxJava.

I've added the requested tests and links, as well as documentation with three examples.

blesh commented:
I've also removed the contents of the dist folder (I think that's what you're asking me to do?)

blesh commented:
@mattpodwysocki looking at it, I'm not sure that you wanted me to remove the files under dist/ .. I guess I'm not sure what you meant by

>  \4. You must remove any files from this commit in the dist folder as it makes it impossible to merge properly.

blesh commented:
I put them back. I *think* this is what you wanted?
